### PR TITLE
Make saga id generation pluggable

### LINKED ISSUE #2445: Make the saga id generation scheme pluggable
Since the current hardcoded scheme is Combguid which is ideal for sqlserver but less ideal for other stores like RavenDB. We also need an option when the user can specify a func of message since that will allow them to use one of the message properties as their saga id. This will allow for a special type of "single correlation id" saga where everything is synced around that one id.

This opens up scenarios like:
- Skipping the unique document in raven and save 1 write and 1 extra read per saga
- Allow us to use the eventstore to store sagas (since it doesn't have cross stream "consistency")

cc @jbogard 

This one depends on:

https://github.com/Particular/NServiceBus/issues/2444

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes #2445 

This enables:

- Simple sagas
- [A filebased development saga persister](https://github.com/Particular/NServiceBus/pull/4496)

Currently the core doesn't make assumptions that all messages need to have a message property which can be always mapped to an existing saga instance. When a new saga instance is created by default a new comb GUID is generated and then messages going out from that saga will contain the saga id header to be able to be correlated back. To ensure partition affinity with Service Fabric we need a deterministic way of creating from a message which started a saga a saga ID based on business identifiers. With that approach it is possible to route messages that need to be sent to an existing saga instance to the endpoint that is living on the partition that contains the saga data.

### CODE REVIEW

File: src/NServiceBus.Core/Sagas/Sagas.cs [hunk 1]
```diff
@@ -43,6 +43,11 @@ protected internal override void Setup(FeatureConfigurationContext context)
                 throw new Exception("The selected persistence doesn't have support for saga storage. Select another persistence or disable the sagas feature using endpointConfiguration.DisableFeature<Sagas>()");
             }
 
+            if (!context.Container.HasComponent<ISagaIdGenerator>())
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Should we use the settings instead to avoid heading further into container hell?

Persisters that wish to change the generator could just do a settings.Set(new MyCustomStrategy())? (is would that mean we have to add a feature dependency)

Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: The persisters actually have the dependency already https://github.com/Particular/NServiceBus.RavenDB/blob/develop/src/NServiceBus.RavenDB/SagaPersister/RavenDbSagaStorage.cs#L13

So until we redesign the seam with a proper "SagaPersistenceDefinition" similar to the transports we should be good

Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Yeah @andreasohlund like I said in the description it should probably something in the persistence definition (could also be a property or like you said added as defaults to the settings)

File: src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs [hunk 2]
```diff
@@ -297,8 +297,14 @@ IContainSagaData CreateNewSagaEntity(SagaMetadata metadata, IInvokeHandlerContex
                     .ConvertFromInvariantString(value.PropertyValue.ToString());
 
                 propertyInfo.SetValue(sagaEntity, convertedValue);
-            }
 
+                sagaEntity.Id = sagaIdGenerator.Generate(value.PropertyName, value.PropertyValue, metadata, context.Extensions);
+            }
+            else
+            {
+                sagaEntity.Id = sagaIdGenerator.Generate(null, null, metadata, context.Extensions);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: I don't really like the approach of passing null values. That will be hard to keep in mind for custom implementations.

File: src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        /// <summary>
+        /// Generates a saga id based on property name and property value.
+        /// </summary>
+        /// <param name="propertyName">The property name. Might be null when a custom finder is used.</param>
+        /// <param name="propertyValue">The property value. Might be null when a custom finder is used.</param>
+        /// <param name="metadata">The saga metadata.</param>
+        /// <param name="context">The context bag.</param>
+        /// <returns>The saga id.</returns>
+        Guid Generate(string propertyName, object propertyValue, SagaMetadata metadata, ContextBag context);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs]: should property be named `correlationProperty` ?

File: src/NServiceBus.Core/Sagas/Sagas.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Looking at the [persitence definition](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Persistence/PersistenceDefinition.cs) again I wonder if we could add this there as a small step towards separated persistence definitions for each relevant storage type.

We could have a optional SagaPersistenceDefiniton GetSagaPersistenceDefintion() on the current PersitenceDefinition where we could add this?

That would make us prepared for a future model where we would get that specific definition from a dedicated factory just like we do with the transports?

File: src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: How about we pass in https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Sagas/SagaCorrelationProperty.cs#L6 which has `.None` (which ends up being `null `as well but anyway ;) )

File: src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs (hunk 3 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs]: See https://github.com/Particular/NServiceBus/pull/4468#discussion_r100485606

File: src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: You're saying to go with a `SagaIdGeneratorContext` or similar? (that would have the added benefit of allowing use to evolve the parameters in a non breaking way)

Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: I think a context would be over the top. how about

`Guid Generate(CorrelationProperty correlationProperty, SagaMetadata metadata, ContextBag context);`

File: src/NServiceBus.Core/Sagas/Sagas.cs [hunk 4]
```diff
@@ -43,6 +43,11 @@ protected internal override void Setup(FeatureConfigurationContext context)
...
+            {
+                context.Container.RegisterSingleton<ISagaIdGenerator>(new DefaultSagaIdGenerator());
```
Comment (WilliamBZA) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: The containers all check for `if HasComponent(componentType)` before registering. The previous check is redundant.
```

File: src/NServiceBus.Core/Sagas/Sagas.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Putting my ServiceFabric TF hat on:

I'm sorry @andreasohlund but we just need saga id generation. Tackling persistence definition in general seems out of scope

File: src/NServiceBus.Core/Sagas/Sagas.cs (hunk 4 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: @WilliamBZA that will most likely go away anyway

Comment (WilliamBZA) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Uhh, actually none of them do the check for singletons. Ignore my previous comment.

File: src/NServiceBus.Core/Sagas/Sagas.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: Wouldn't it be very little code to make it explicit instead of having a black magic `setting.Set<ISagaIdGenerator>(new ServiceFabricAwareIdGenerator())`?

But I do see the YAGNI aspect to this so I'm fine either way

File: src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: Discussed with @timbussmann and @andreasohlund will add a context hear as well but we won't make any changes to the persistence definition

**Comment by danielmarbach** [comment]
We discovered that we don't actually need this as of now. Happy to close the PR if we don't want to tackle it further.

**Comment by bording** [comment]
@danielmarbach If you don't actually need this for the Service Fabric stuff, what's the use case that this enables, and is it worth making changes now to enable it?

If not, I'd say go ahead and close the PR.

**Comment by danielmarbach** [comment]
Talked to @andreasohlund and we still see a lot of value in this. Assigned it to @andreasohlund since he is going to explore the impact on this to the file based persistence

**Comment by andreasohlund** [comment]
I've confirmed that this work and does unlock persisters to make use of the single correlated sagas idea, see https://github.com/Particular/NServiceBus/pull/4496

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers ok if I go ahead and include the settings approach

https://github.com/Particular/NServiceBus/pull/4496/files#diff-46678676e33293f7ab3bfa0cbf703457R14
https://github.com/Particular/NServiceBus/pull/4496/files#diff-e3b2dc6292805e519102015af1d0dc26R46

in this PR so that we can pull this into 6.2?

**Comment by timbussmann** [comment]
@andreasohlund as this is likely something used by downstreams, can we at least provide a public API which wraps the settings approach in order to point out, that this might indeed by used by downstreams? (not that we remove that accidentally later)

**Comment by andreasohlund** [comment]
Decided to skip a explicit public api for now since this is only to be used by extension points and we'll likely go with a explicit `SagaPersistenceDefinition` in the future.

```
var definition = context.Settings.Get<SagaPersistenceDefintions>();

var idGenerator = definition.GetSagaIdGenerator();

```

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers  Changes mentioned in https://github.com/Particular/NServiceBus/pull/4468#issuecomment-283051382 has been completed. This one is ready for final review

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 5]
```diff
@@ -2745,6 +2749,13 @@ namespace NServiceBus.Sagas
         public System.Collections.Generic.Dictionary<string, object> Properties { get; }
         public System.Type Type { get; }
     }
+    public class SagaIdGeneratorContext : NServiceBus.Extensibility.ContextBag
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Normally we have a context property on the context and do not inherit from the context bag so that you have to explicitly do context.Context.TryGet... I know it is a bit more verbose but it hides the extensibility part and doesn't expose it it "in your face style" in intellisense

Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: Fixed

File: src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs [hunk 6]
```diff
@@ -5,7 +5,7 @@ namespace NServiceBus.Sagas
     /// <summary>
     /// Context provided to the saga id generator.
     /// </summary>
-    public class SagaIdGeneratorContext : ContextBag
+    public class SagaIdGeneratorContext : IExtendable
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: Interesting, we haven't done this with other context. Should we align others or remove the interface here?

https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Routing/DistributionContext.cs#L49

Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: I'n good either way, just used on some other context and copy pasted it

Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: Does anyone remeber the original intent of that interface? I like the fact that it makes the naming consistent (to avoid the ugly context.Context we have on a few other contexts)

**Comment by danielmarbach** [comment]
@andreasohlund I cannot approve this since it is my PR ;)

**Comment by bording** [comment]
This has conflicts that need to be resolved.

**Comment by andreasohlund** [comment]
@bording rebased

File: src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs [hunk 7]
```diff
@@ -329,8 +331,16 @@ IContainSagaData CreateNewSagaEntity(SagaMetadata metadata, IInvokeHandlerContex
...
                 propertyInfo.SetValue(sagaEntity, convertedValue);
+
+                sagaIdGeneratorContext = new SagaIdGeneratorContext(value.PropertyName, value.PropertyValue, metadata, context.Extensions);
+            }
+            else
+            {
+                sagaIdGeneratorContext = new SagaIdGeneratorContext(null, null, metadata, context.Extensions);
             }
 
+            sagaEntity.Id = sagaIdGenerator.Generate(sagaIdGeneratorContext);
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs]: This could be simplified by doing this instead:

```csharp
var sagaIdGeneratorContext = new SagaIdGeneratorContext(value?.PropertyName, value?.PropertyValue, metadata, context.Extensions);
```

File: src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    /// <summary>
+    /// Context provided to the saga id generator.
+    /// </summary>
+    public class SagaIdGeneratorContext : IExtendable
+    {
+
+        /// <summary>
+        /// Constructs a new context.
+        /// </summary>
+        /// <param name="correlationPropertyName">Name of property used to correlate messages to this saga. Can be <code>null</code> if a custom saga finder is used for the given message.</param>
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: "Name of **the** property"

File: src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            CorrelationPropertyName = correlationPropertyName;
+            CorrelationPropertyValue = correlationPropertyValue;
+            SagaMetadata = sagaMetadata;
+            Extensions = extensions;
+        }
+
+        /// <summary>
+        /// Name of property used to correlate messages to this saga.
+        /// </summary>
+        public string CorrelationPropertyName { get; private set; }
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: These properties can all be get-only.

File: src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            Guard.AgainstNull(nameof(extensions), extensions);
+
+            CorrelationPropertyName = correlationPropertyName;
+            CorrelationPropertyValue = correlationPropertyValue;
+            SagaMetadata = sagaMetadata;
+            Extensions = extensions;
+        }
+
+        /// <summary>
+        /// Name of property used to correlate messages to this saga.
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: "Name of **the** property"

File: src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+{
+    using System;
+
+    /// <summary>
+    /// Saga id generator.
+    /// </summary>
+    public interface ISagaIdGenerator
+    {
+        /// <summary>
+        /// Generates a saga id based on property name and property value.
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs]: Should this be "based on the correlation property name and value" instead?

File: src/NServiceBus.Core/Extensibility/ContextBag.cs [hunk 12]
```diff
@@ -10,9 +10,9 @@ public class ContextBag : ReadOnlyContextBag
         /// <summary>
         /// Initialized a new instance of <see cref="ContextBag" />.
         /// </summary>
-        public ContextBag(ContextBag parentBag = null)
+        public ContextBag(ContextBag parentContext = null)
```
Comment (bording) [inline src/NServiceBus.Core/Extensibility/ContextBag.cs]: If we're changing the name of the parameter/field, perhaps `parent` would be a better name, since Bag/Context is a bit redundant?

File: src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Metadata for the targeted saga.
+        /// </summary>
+        public SagaMetadata SagaMetadata { get; private set; }
+
+        /// <summary>
+        /// A <see cref="ContextBag" /> which can be used to extend the current object.
+        /// </summary>
+        public ContextBag Extensions { get; }
```
Comment (bording) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: What's the use case around having this ContextBag available here? Who would be able to add things to it?

Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/SagaIdGeneratorContext.cs]: We tend to include it on all contexts we have "just in case" since it allows us/users to provide more things without breaking changes, happy to reeveluate that decision? ( @Particular/nservicebus-maintainers )

File: src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs (hunk 11 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/ISagaIdGenerator.cs]: I just realized that property name and value is optional (if a custom finder is used) so how about:

> Generates a custom saga id

?

**Comment by andreasohlund** [comment]
@bording https://github.com/Particular/NServiceBus/pull/4468/commits/2df767baf7b686414541f84f2b9e39c08ccf3385 should address your comments. Note how I switched to the `SagaCorrelationProperty` that IMO is a better api since it allows persisters (like the dev persister) that doesn't support custom finders to throw if the generator is called with a non existing correlation property.

`if(correlationProperty == SagaCorrelationProperty.None) throw new NotSupportedException("bla")`

Thoughts?

**Comment by bording** [comment]
LGTM, though I'd like to see what everyone else thinks regarding the Extensions ContextBag.

**Comment by danielmarbach** [comment]
@bording so far if something is called in the pipeline we added it. It is the only way to float in pipeline dependencies like a connection or something similar into the things that are called from within the pipeline.  I have a hunch that we will regret it if we don't add it ;)

**Comment by andreasohlund** [comment]
I will proceed with a doco PR for this
